pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/gates/TNCudaGateCache.hpp

```diff
@@ -0,0 +1,139 @@
+
+    ~TNCudaGateCache(){};
+
+    /**
+     * @brief Add gate numerical value to the cache, indexed by the id of gate
+     * tensor operator in the graph and its name and parameter value are
+     * recorded as well.
+     *
+     * @param gate_id The id of gate tensor operator in the computate graph.
+     * @param gate_name String representing the name of the given gate.
+     * @param gate_param Gate parameter value. `0.0` if non-parametric gate.
+     */
+    void add_gate(const std::size_t gate_id, const std::string &gate_name,
+                  [[maybe_unused]] std::vector<PrecisionT> gate_param) {
+        auto gate_key = std::make_pair(gate_name, gate_param);
+
+        auto &gateMap =
+            cuGates::DynamicGateDataAccess<PrecisionT>::getInstance();
+
+        add_gate(gate_id, gate_key, gateMap.getGateData(gate_name, gate_param));
+    }
+    /**
+     * @brief Add gate numerical value to the cache, indexed by the id of gate
+     * tensor operator in the graph and its name and parameter value as well as
+     * the gate data on host.
+     *
+     * @param gate_id The id of gate tensor operator in the computate graph.
+     * @param gate_key String representing the name of the given gate as well as
+     * its associated parameter value.
+     * @param gate_data_host Vector of complex floating point values
+     * representing the gate data on host.
+     */
+
+    void add_gate(const std::size_t gate_id, gate_key_info gate_key,
+                  const std::vector<CFP_t> &gate_data_host) {
+        std::size_t rank = Pennylane::Util::log2(gate_data_host.size());
+        auto modes = std::vector<std::size_t>(rank, 0);
+        auto extents = std::vector<std::size_t>(rank, 2);
+
+        auto &&tensor =
+            TensorCuda<PrecisionT>(rank, modes, extents, device_tag_);
+
+        device_gates_.emplace(
+            std::piecewise_construct, std::forward_as_tuple(gate_id),
+            std::forward_as_tuple(gate_key, std::move(tensor)));
+
+        device_gates_.at(gate_id).second.getDataBuffer().CopyHostDataToGpu(
+            gate_data_host.data(), gate_data_host.size());
+
+        total_alloc_bytes_ += (sizeof(CFP_t) * gate_data_host.size());
+    }
+
+    /**
+     * @brief Returns a pointer to the GPU device memory where the gate is
+     * stored.
+     *
+     * @param gate_id The id of gate tensor operator in the computate graph.
+     * @return const CFP_t* Pointer to gate values on device.
+     */
+    CFP_t *get_gate_device_ptr(const std::size_t gate_id) {
+        return device_gates_.at(gate_id).second.getDataBuffer().getData();
+    }
+
+  private:
+    const DevTag<int> device_tag_;
+    std::size_t total_alloc_bytes_;
+
+    struct gate_info_hash {
+        std::size_t operator()(std::size_t key_id) const {
+            return std::hash<std::size_t>()(key_id);
+        }
+    };
+
+    std::unordered_map<std::size_t, gate_info, gate_info_hash> device_gates_;
```

Comment:
Would you be able to add a brief explanation for `device_gates_` and use cases as per our offline conversations?

Reply:
Good point! Done!